### PR TITLE
[Dubbo-3305] Fix performance overhead caused by too much unknown classloading during hessian2 deserialization

### PR DESCRIPTION
**Pull request by hopelove404**
## What is the purpose of the change

To avoid frequently class loading and to reduce performance overhead during the process of hessian2 deserialization for those classes are unknown in current classloader

## Brief changelog

override method public Deserializer getDeserializer(String type) in Hessian2SerializerFactory.java
 refactor get hessian2 deserializer for string type
## Verifying this change

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/incubator-dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/incubator-dubbo/tree/master/dubbo-test).
- [x] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/incubator-dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-serialization/dubbo-serialization-hessian2/src/main/java/org/apache/dubbo/common/serialize/hessian2/Hessian2SerializerFactory.java [hunk 1]
```diff
@@ -30,4 +44,18 @@ public ClassLoader getClassLoader() {
...
+    @Override
+    public Deserializer getDeserializer(String type)
+            throws HessianProtocolException {
+        if (type == null || type.equals("") || typeNotFoundDeserializer.contains(type)) {
+            return null;
+        }
+        Deserializer deserializer = super.getDeserializer(type);
+        if (deserializer == null) {
+            typeNotFoundDeserializer.add(type);
+            logger.warn("unable to find deserializer for string type:" + type);
```
Comment (CrazyHZM) [inline dubbo-serialization/dubbo-serialization-hessian2/src/main/java/org/apache/dubbo/common/serialize/hessian2/Hessian2SerializerFactory.java]: `logger.warn("Failed to find deserializer with type:" + type);`

File: dubbo-serialization/dubbo-serialization-hessian2/src/main/java/org/apache/dubbo/common/serialize/hessian2/Hessian2SerializerFactory.java [hunk 2]
```diff
@@ -30,4 +44,18 @@ public ClassLoader getClassLoader() {
         return Thread.currentThread().getContextClassLoader();
     }
 
+    @Override
+    public Deserializer getDeserializer(String type)
+            throws HessianProtocolException {
+        if (type == null || type.equals("") || typeNotFoundDeserializer.contains(type)) {
```
[the remark this answers is not in the document]
Comment (hopelove404) [reply dubbo-serialization/dubbo-serialization-hessian2/src/main/java/org/apache/dubbo/common/serialize/hessian2/Hessian2SerializerFactory.java]: thanks，already replaced with StringUtils#isEmpty

**Comment by lovepoem** [comment]
1. Can this change be moved to https://github.com/dubbo/hessian-lite? 
2. Any code use the `typeNotFoundDeserializer ` ? 
3. Please add some test case to verify your pr

**Comment by hopelove404** [comment]
@lovepoem 
> 1. Can this change be moved to https://github.com/dubbo/hessian-lite?
> thanks for your advice,I think it would be better to move this piece of code 
 to https://github.com/dubbo/hessian-lite. Do I need to open a new pr in hessian-lite and close this one?

> 2. Any code use the `typeNotFoundDeserializer ` ?
>typeNotFoundDeserializer is only used in this class.

> 3. Please add some test case to verify your pr
>I will add the tests later.

**Comment by lovepoem** [comment]
> thanks for your advice,I think it would be better to move this piece of
code
to https://github.com/dubbo/hessian-lite. Do I need to open a new pr in
hessian-lite and close this one?

If so , please work on the repository :
https://github.com/dubbo/hessian-lite  and test is

hopelove404 <notifications@github.com> 于2019年1月28日周一 下午6:41写道：

**Comment by beiwei30** [comment]
@hopelove404, I agree with @lovepoem's suggestion. This blacklist logic should put into **dubbo/hessian-lite** instead of dubbo project.

File: dubbo-serialization/dubbo-serialization-hessian2/src/main/java/org/apache/dubbo/common/serialize/hessian2/Hessian2SerializerFactory.java [hunk 3]
```diff
@@ -16,12 +16,27 @@
...
 public class Hessian2SerializerFactory extends SerializerFactory {
+    private final static Logger logger = LoggerFactory.getLogger(Hessian2SerializerFactory.class);
 
     public static final SerializerFactory SERIALIZER_FACTORY = new Hessian2SerializerFactory();
 
+    /**
+     * For those classes are unknown in current classloader, record them in this set to avoid
+     * frequently class loading and to reduce performance overhead.
+     */
+    private Set<String> typeNotFoundDeserializer = new ConcurrentHashSet<String>(8);
```
Comment (khanimteyaz) [inline dubbo-serialization/dubbo-serialization-hessian2/src/main/java/org/apache/dubbo/common/serialize/hessian2/Hessian2SerializerFactory.java]: You can use the dimond operator.

private Set<String> typeNotFoundDeserializer = new ConcurrentHashSet<>(8);

Comment (khanimteyaz) [reply dubbo-serialization/dubbo-serialization-hessian2/src/main/java/org/apache/dubbo/common/serialize/hessian2/Hessian2SerializerFactory.java]: Would you care explain us about using 8 as size for ConcurrentHashSet?

**Comment by hopelove404** [comment]
@lovepoem @beiwei30 According to your advice, the above logic has been move to hessian-lite, please help check the new pr. https://github.com/dubbo/hessian-lite/pull/7

**Comment by beiwei30** [comment]
> @lovepoem @beiwei30 According to your advice, the above logic has been move to hessian-lite, please help check the new pr. [dubbo/hessian-lite#7](https://github.com/dubbo/hessian-lite/pull/7)

Sure, I will do it.

**Pull request closed without merging**